registry/repository.go

```diff
@@ -1,198 +1,117 @@
+// artifact manifest descriptor
+func (c *repositoryClient) ListSignatures(ctx context.Context, desc ocispec.Descriptor, fn func(signatureManifests []ocispec.Descriptor) error) error {
+	refFinder, ok := c.Repository.(orasRegistry.ReferrerFinder)
+	if !ok {
+		return errors.New("repo is not a orasRegistry.ReferrerFinder")
 	}
-	return signatureManifests, nil
+	return refFinder.Referrers(ctx, desc, ArtifactTypeNotation, fn)
 }
 
-// GetBlob downloads the content of the specified digest's Blob
-func (c *RepositoryClient) GetBlob(ctx context.Context, digest digest.Digest) ([]byte, error) {
-	desc, err := c.Repository.Blobs().Resolve(ctx, digest.String())
+// FetchSignatureBlob returns signature envelope blob and descriptor given
+// signature manifest descriptor
+func (c *repositoryClient) FetchSignatureBlob(ctx context.Context, desc ocispec.Descriptor) ([]byte, ocispec.Descriptor, error) {
+	sigManifest, err := c.getSignatureManifest(ctx, desc)
 	if err != nil {
-		return nil, err
+		return nil, ocispec.Descriptor{}, err
 	}
-	if desc.Size > maxBlobSizeLimit {
-		return nil, fmt.Errorf("signature blob too large: %d", desc.Size)
+	if len(sigManifest.Blobs) != 1 {
+		return nil, ocispec.Descriptor{}, fmt.Errorf("signature manifest requries exactly one signature envelope blob, got %d", len(sigManifest.Blobs))
 	}
-	return content.FetchAll(ctx, c.Repository.Blobs(), desc)
+	sigDesc := sigManifest.Blobs[0]
+	if sigDesc.Size > maxBlobSizeLimit {
+		return nil, ocispec.Descriptor{}, fmt.Errorf("signature blob too large: %d", sigDesc.Size)
+	}
+	sigBlob, err := content.FetchAll(ctx, c.Repository.Blobs(), sigDesc)
+	if err != nil {
+		return nil, ocispec.Descriptor{}, err
+	}
+	return sigBlob, sigDesc, nil
 }
 
-// PutSignatureManifest creates and uploads an signature artifact linking the manifest and the signature
-func (c *RepositoryClient) PutSignatureManifest(ctx context.Context, signature []byte, signatureMediaType string, subjectManifest notation.Descriptor, annotations map[string]string) (notation.Descriptor, SignatureManifest, error) {
-	signatureDesc, err := c.uploadSignature(ctx, signature, signatureMediaType)
+// PushSignature creates and uploads an signature manifest along with its
+// linked signature envelope blob. Upon successful, PushSignature returns
+// signature envelope blob and manifest descriptors.
+func (c *repositoryClient) PushSignature(ctx context.Context, blob []byte, mediaType string, subject ocispec.Descriptor, annotations map[string]string) (blobDesc, manifestDesc ocispec.Descriptor, err error) {
+	blobDesc, err = oras.PushBytes(ctx, c.Repository.Blobs(), mediaType, blob)
 	if err != nil {
-		return notation.Descriptor{}, SignatureManifest{}, err
+		return ocispec.Descriptor{}, ocispec.Descriptor{}, err
 	}
 
-	manifestDesc, err := c.uploadSignatureManifest(ctx, artifactDescriptorFromNotation(subjectManifest), signatureDesc, annotations)
+	manifestDesc, err = c.uploadSignatureManifest(ctx, subject, blobDesc, annotations)
```

Comment:
Should we add validation around manifest mediaType?

Reply:
Added validation.